mobile-browser-based-version/src/logic/communication/federated/client_async.ts

```diff
@@ -0,0 +1,140 @@
+import * as api from './api'
+
+/**
+ * Class that deals with communication with the centralized server when training
+ * a specific task.
+ */
+export class FederatedAsyncClient extends Client {
+  clientID: string;
+  peer: any;
+  round: number;
+
+  /**
+   * Prepares connection to a centralized server for training a given task.
+   * @param {String} serverURL The URL of the centralized server.
+   * @param {Task} task The associated task object.
+   * @param {Number} round The training round.
+   */
+  constructor (serverURL, task) {
+    super(serverURL, task)
+    console.log('building Federated Async client')
+    this.clientID = ''
+    this.round = -1 // The server starts at round 0, in the beginning we are behind
+  }
+
+  /**
+   * Initialize the connection to the server. TODO: In the case of FeAI,
+   * should return the current server-side round for the task.
+   */
+  async connect () {
+    /**
+     * Create an ID used to connect to the server.
+     * The client is now considered as connected and further
+     * API requests may be made.
+     */
+    this.clientID = makeID(10)
+    const response = await api.connect(this.task.taskID, this.clientID)
+    return response.status === 200
+  }
+
+  /**
+   * Disconnection process when user quits the task.
+   */
+  async disconnect () {
+    const response = await api.disconnect(this.task.taskID, this.clientID)
+    return response.status === 200
+  }
+
+  async postWeights (weights) {
+    const encodedWeights = msgpack.encode(
+      Array.from(await serializeWeights(weights))
+    )
+    const response = await api.postAsyncWeights(
+      this.task.taskID,
+      this.clientID,
+      encodedWeights,
+      this.round
+    )
+    return response.status === 200
+  }
+
+  async postMetadata (metadataID, metadata) {
+    const response = api.postMetadata(
+      this.task.taskID,
+      this.round,
+      this.clientID,
+      metadataID,
+      metadata
+    )
+    return (await response).status === 200
+  }
+
+  async getMetadataMap (metadataID) {
+    const response = await api.getMetadataMap(
+      this.task.taskID,
+      this.round,
+      this.clientID,
+      metadataID
+    )
+    if (response.status === 200) {
+      const body = await response.data
+      return new Map(msgpack.decode(body[metadataID]))
+    } else {
+      return new Map()
+    }
+  }
+
+  async _getServerRound (): Promise<number> {
+    const response = await api.getAsyncRound(this.task.taskID, this.clientID)
+
+    if (response.status === 200) {
+      return response.data.round
+    }
+    console.log('Error getting weights: code', response.status)
+    return -1
+  }
+
+  _updateLocalModelWithMostRecentServerModel () {
+  // TODO: naming is not good, seems like you are just creating + only do so if it is more recent.
+  //! !!!!
+    this.task.createModel()
+    console.log('Updated local model')
+  }
+
+  async onEpochBeginCommunication (model, epoch, trainingInformant) {
```

Comment:
why duplicate the `_update` call from `onEpochEndCommunication`? the two callbacks are triggered very close in time, so I don't think it would be beneficial to make more frequent calls to `_update` in this way

Reply:
I basically tried to copy what was in the sync client. Does it make more sense there to update twice? 
For the async, is it better than to just keep it at epoch begin? (since we need to load the weights, or maybe this is already done before?)